File: java/OOAD/TicTacToe/src/com/techlabs/tictactoe/ResultAnalyzer.java
package com.techlabs.tictactoe;

public class ResultAnalyzer {
	private Board board;
	private Cell[][] cells;
	private GameStatus gameStatus;

	public ResultAnalyzer(Board board) {
		this.board = board;
		this.cells = this.board.getCells();
	}

	public GameStatus generateResult(Board board) {
		this.board = board;
		boolean result;

		for (int i = 0; i < board.getNoRows(); i++) {
			for (int j = 0; j < board.getNoColumns(); j++) {
				if (!(cells[i][j].getCellMark() == Mark.N)) {
					// analyze the cell
					result = analyzeCell(cells[i][j].getPostion());
					if (result) {
						return this.gameStatus = GameStatus.WON;
						
					} else {
						this.gameStatus = GameStatus.ONGOING;
					}
				}
			}
		}
		
		if (checkDrawCase() && (this.gameStatus==GameStatus.ONGOING)) {
			this.gameStatus = GameStatus.DRAW;
		}

		return this.gameStatus;
	}

	private boolean analyzeCell(int position) {
		switch (position) {
		case 1:
			if (checkHorizontalWin(0) || checkVerticalWin(0)
					|| checkDiagonalWin()) {
				return true;
			}
			break;

		case 2:
			if (checkHorizontalWin(0) || checkVerticalWin(1)) {
				return true;
			}
			break;
		case 3:
			if (checkHorizontalWin(0) || checkVerticalWin(2)
					|| checkDiagonalWin()) {
				return true;
			}
			break;
		case 4:
			if (checkHorizontalWin(1) || checkVerticalWin(0)) {
				return true;
			}
			break;
		case 5:
			if (checkHorizontalWin(1) || checkVerticalWin(1)
					|| checkDiagonalWin()) {
				return true;
			}
			break;
		case 6:
			if (checkHorizontalWin(1) || checkVerticalWin(2)) {
				return true;
			}
			break;
		case 7:
			if (checkHorizontalWin(2) || checkVerticalWin(0)
					|| checkDiagonalWin()) {
				return true;
			}
			break;
		case 8:
			if (checkHorizontalWin(2) || checkVerticalWin(1)) {
				return true;
			}
			break;
		case 9:
			if (checkHorizontalWin(2) || checkVerticalWin(2)
					|| checkDiagonalWin()) {
				return true;
			}
			break;
		default:
			return false;

		}
		return false;

	}

	public boolean checkDrawCase() {

		if (this.gameStatus == GameStatus.ONGOING) {
			for (int i = 0; i < board.getNoRows(); i++) {
				for (int j = 0; j < board.getNoColumns(); j++) {
					if (cells[i][j].getCellMark() == Mark.N) {
						return false;
						
					}
				}
			}
		}
		return true;
	}

	private boolean checkDiagonalWin() {

		Mark markAtZeroZeroPosition = cells[0][0].getCellMark();
		Mark markAtTwoZeroPos = cells[2][0].getCellMark();
		if (markAtZeroZeroPosition == cells[1][1].getCellMark()
				&& markAtZeroZeroPosition == cells[2][2].getCellMark()) {
			if (markAtZeroZeroPosition != Mark.N
					&& cells[1][1].getCellMark() != Mark.N
					&& cells[2][2].getCellMark() != Mark.N) {
				return true;
			} else {
				return false;
			}

		} else if (markAtTwoZeroPos == cells[1][1].getCellMark()
				&& markAtTwoZeroPos == cells[0][2].getCellMark()) {
			if (markAtTwoZeroPos != Mark.N
					&& cells[1][1].getCellMark() != Mark.N
					&& cells[0][2].getCellMark() != Mark.N) {
				return true;
			} else {
				return false;
			}

		}
		return false;

	}

	private boolean checkVerticalWin(int col) {

		if (cells[0][col].getCellMark() == cells[1][col].getCellMark()) {
			if (cells[0][col].getCellMark() == cells[2][col].getCellMark()) {
				return true;
			}
		}
		return false;
	}

	private boolean checkHorizontalWin(int row) {
		if (cells[row][0].getCellMark() == cells[row][1].getCellMark()) {
			if (cells[row][0].getCellMark() == cells[row][2].getCellMark()) {
				return true;
			}
		}
		return false;
	}

}

File: dotNet/classlibrary/WinForm/TicTacToeApp/TicTacToeCore/ResultAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TicTacToeCore
{
   public class ResultAnalyzer
    {
        private Board _board;
        private Cell[,] cells;
        private GameStatus _gameStatus;

        public ResultAnalyzer(Board board)
        {
            this._board = board;
            this.cells = this._board.Cells;
        }

        public GameStatus GenerateResult(Board board)
        {
            this._board = board;
            bool result;
            for (int i = 0; i < _board.NoRows; i++)
            {
                for (int j = 0; j < _board.NoCols; j++)
                {
                    if (!(cells[i,j].CellMark.Mark == MarkFactory.NoValue.Mark))
                    {
                        // analyze the cell
                        result = AnalyzeCell(cells[i,j].Position);
                        if (result)
                        {
                            return this._gameStatus = GameStatus.WON;

                        }
                        else
                        {
                            this._gameStatus = GameStatus.ONGOING;
                        }
                    }
                }
            }

            if (CheckDrawCase() && (this._gameStatus == GameStatus.ONGOING))
            {
                this._gameStatus = GameStatus.DRAW;
            }

            return this._gameStatus;
        }

        private bool AnalyzeCell(int position)
        {
            switch (position)
            {
                case 1:
                    if (CheckHorizontalWin(0) || CheckVerticalWin(0)
                            || CheckDiagonalWin())
                    {
                        return true;
                    }
                    break;

                case 2:
                    if (CheckHorizontalWin(0) || CheckVerticalWin(1))
                    {
                        return true;
                    }
                    break;
                case 3:
                    if (CheckHorizontalWin(0) || CheckVerticalWin(2)
                            || CheckDiagonalWin())
                    {
                        return true;
                    }
                    break;
                case 4:
                    if (CheckHorizontalWin(1) || CheckVerticalWin(0))
                    {
                        return true;
                    }
                    break;
                case 5:
                    if (CheckHorizontalWin(1) || CheckVerticalWin(1)
                            || CheckDiagonalWin())
                    {
                        return true;
                    }
                    break;
                case 6:
                    if (CheckHorizontalWin(1) || CheckVerticalWin(2))
                    {
                        return true;
                    }
                    break;
                case 7:
                    if (CheckHorizontalWin(2) || CheckVerticalWin(0)
                            || CheckDiagonalWin())
                    {
                        return true;
                    }
                    break;
                case 8:
                    if (CheckHorizontalWin(2) || CheckVerticalWin(1))
                    {
                        return true;
                    }
                    break;
                case 9:
                    if (CheckHorizontalWin(2) || CheckVerticalWin(2)
                            || CheckDiagonalWin())
                    {
                        return true;
                    }
                    break;
                default:
                    return false;

            }
            return false;

        }

        public bool CheckDrawCase()
        {

            if (this._gameStatus == GameStatus.ONGOING)
            {
                for (int i = 0; i < _board.NoRows; i++)
                {
                    for (int j = 0; j < _board.NoCols; j++)
                    {
                        if (cells[i,j].CellMark.Mark == MarkFactory.NoValue.Mark)
                        {
                            return false;

                        }
                    }
                }
            }
            return true;
        }

        private bool CheckDiagonalWin()
        {

            MarkType markAtZeroZeroPosition = cells[0,0].CellMark;
            MarkType markAtTwoZeroPos = cells[2,0].CellMark;
            if (markAtZeroZeroPosition.Mark == cells[1,1].CellMark.Mark
                    && markAtZeroZeroPosition.Mark == cells[2,2].CellMark.Mark)
            {
                if (markAtZeroZeroPosition.Mark != MarkFactory.NoValue.Mark
                        && cells[1,1].CellMark.Mark != MarkFactory.NoValue.Mark
                        && cells[2,2].CellMark.Mark != MarkFactory.NoValue.Mark)
                {
                    return true;
                }
                else
                {
                    return false;
                }

            }
            else if (markAtTwoZeroPos.Mark == cells[1,1].CellMark.Mark
                  && markAtTwoZeroPos.Mark == cells[0,2].CellMark.Mark)
            {
                if (markAtTwoZeroPos.Mark != MarkFactory.NoValue.Mark
                        && cells[1,1].CellMark.Mark != MarkFactory.NoValue.Mark
                        && cells[0,2].CellMark.Mark != MarkFactory.NoValue.Mark)
                {
                    return true;
                }
                else
                {
                    return false;
                }

            }
            return false;

        }

        private bool CheckVerticalWin(int col)
        {

            if (cells[0,col].CellMark.Mark == cells[1,col].CellMark.Mark)
            {
                if (cells[0,col].CellMark.Mark == cells[2,col].CellMark.Mark)
                {
                    return true;
                }
            }
            return false;
        }

        private bool CheckHorizontalWin(int row)
        {
            if (cells[row,0].CellMark.Mark == cells[row,1].CellMark.Mark)
            {
                if (cells[row,0].CellMark.Mark == cells[row,2].CellMark.Mark)
                {
                    return true;
                }
            }
            return false;
        }





    }

}
